File: aoc-py/solutions/day8.py
"""
Day 8: Playground

https://adventofcode.com/2025/day/8
"""
from __future__ import annotations

__all__ = ('Day8',)

from typing import ClassVar

from math import dist
from collections import defaultdict
from itertools import combinations

from ..solution import Solution

class DisjointSet[T]:
    def __init__(self, nodes: list[T]) -> None:
        self._nodes = nodes
        # we will use a tree to represent a set
        #
        # maps `node` -> the direct parent of `node`
        # start off with all nodes being disjoint sets (sets containing just itself, so its parent will also be itself)
        # i.e. given nodes [1, 2, 3] we will re-organize them with {1:1, 2:2, 3:3}
        self._parents = {box: box for box in nodes}

    def root(self, node: T) -> T:
        """Finds the root node of the set containing `node`, a unique identifier representing a disjoint set"""
        # if the current node's parent is itself
        # then we are at the root just like how we defined it in the constructor
        if self._parents[node] == node:
            return node

        # optimization: path compression
        # if we find that our tree is: 3->4->5->...->root
        # we can just re-declare 3's parent to be 5 (the parent of 4 (the parent of 3))
        # eventually, due to recursion, the path will compress it such that all nodes (i.e. 3) will be connected to the root directly
        # reducing the depth of the tree (since `root` is recursive and is slower if the tree is deeper)
        self._parents[node] = self.root(self._parents[node])
        # now that our parent is our root, we can just return our parent
        return self._parents[node]

    def get_sizes(self) -> dict[T, int]:
        """Returns a mapping of `root each set` -> the size of the set

        since a set's root is a unique identifier for the set
        """
        sizes = defaultdict(int)

        for node in self._nodes:
            sizes[self.root(node)] += 1
        return sizes

    def count(self) -> int:
        """Returns the number of disjoint sets in this structure"""

        # check the size of the set containing the roots (identifier) for each disjoint set
        return len({self.root(node) for node in self._nodes})

    def union(self, node1: T, node2: T) -> None:
        """Unions (merges) 2 disjoint sets together (into 1 "joint" set)

        Specifically, we are merging the set containing `node2` into the set containing `node1`
        """
        # find the root of `node2`, and make it a child of the root of `node1`
        # i.e. declare that the parent of the root of `node2` is the root of `node1`
        self._parents[self.root(node2)] = self.root(node1)

class Day8(Solution):
    NAME: ClassVar[str] = 'Playground'

    def part_one(self, inp: str) -> int:
        boxes = [
            tuple(int(a) for a in box.split(','))
            for box in inp.splitlines()
        ]
        sets = DisjointSet(boxes)
        # find all possible pairs of boxes, and sort them by their distance (edge length, low->high)
        edges = sorted(combinations(boxes, 2), key=lambda edge: dist(edge[0], edge[1]))

        # process the 1000 shortest edges
        for box1, box2 in edges[:1000]:
            sets.union(box1, box2)

        sizes = sorted(sets.get_sizes().values(), reverse=True)
        # product of the sizes of the 3 largest sets
        return sizes[0] * sizes[1] * sizes[2]

    def part_two(self, inp: str) -> int:
        boxes = [
            tuple(int(a) for a in box.split(','))
            for box in inp.splitlines()
        ]
        sets = DisjointSet(boxes)
        # find all possible pairs of boxes, and sort them by their distance (edge length, low->high)
        edges = sorted(combinations(boxes, 2), key=lambda edge: dist(edge[0], edge[1]))

        last_box1 = None
        last_box2 = None

        for box1, box2 in edges:
            # process edges until everything is inside 1 connected set (tree)
            if sets.count() == 1:
                break
            # record the last pair of boxes that were processed before we stop
            last_box1 = box1
            last_box2 = box2
            sets.union(box1, box2)

        assert last_box1 and last_box2
        return last_box1[0] * last_box2[0]

    def run(self, inp: str) -> None:
        print('Part 1:', p1 := self.part_one(inp))
        print('Part 2:', p2 := self.part_two(inp))

        assert p1 == 54180
        assert p2 == 25325968

File: src/bin/day8.rs
//! Day 8: Playground
//!
//! <https://adventofcode.com/2025/day/8>
#![allow(future_incompatible)]

use std::{
    collections::{HashSet, HashMap},
    hash::Hash,
};
use itertools::Itertools;
use aoc_2025::Solution;

struct DisjointSet<'a, T>
where
    T: Eq + Hash,
{
    nodes: &'a [T],
    parents: HashMap<&'a T, &'a T>,
}

impl<'a, T> DisjointSet<'a, T>
where
    T: Eq + Hash,
{
    #[must_use]
    pub fn new(nodes: &'a [T]) -> Self {
        Self {
            nodes,
            parents: nodes
                .iter()
                .map(|node| (node, node))
                .collect::<HashMap<_, _>>()
        }
    }

    #[must_use]
    pub fn root(&mut self, node: &'a T) -> &'a T {
        if self.parents[node] == node {
            return node;
        }

        let root = self.root(self.parents[node]);
        self.parents.insert(node, root);
        root
    }

    #[must_use]
    pub fn get_sizes(&mut self) -> HashMap<&'a T, usize> {
        let mut sizes = HashMap::with_capacity(self.parents.len());

        for node in self.nodes {
            *sizes.entry(self.root(node))
                .or_insert(0) += 1;
        }
        sizes
    }

    #[must_use]
    pub fn count(&mut self) -> usize {
        self.nodes
            .iter()
            .map(|node| self.root(node)
        )
        .collect::<HashSet<_>>()
        .len()
    }

    pub fn union(&mut self, node1: &'a T, node2: &'a T) {
        let root1 = self.root(node1);
        let root2 = self.root(node2);

        self.parents.insert(root2, root1);
    }
}

type BoxT = [usize; 3];

pub struct Day8;

impl Day8 {
    #[must_use]
    #[inline]
    fn parse_input(inp: &str) -> (Vec<BoxT>, Vec<(BoxT, BoxT)>) {
        let boxes = inp
            .lines()
            .map(|line| {
                line
                    .split(',')
                    .map(|c| c.parse().unwrap())
                    .collect_array()
                    .unwrap()
            })
            .collect::<Vec<BoxT>>();

        let edges = boxes
            .iter()
            .copied()
            .tuple_combinations()
            .sorted_unstable_by_key(|(a, b)| a
                .iter()
                .zip(b.iter())
                .map(|(&x1, &x2)| x1.abs_diff(x2).pow(2))
                .sum::<usize>()
            )
            .collect::<Vec<_>>();

        (boxes, edges)
    }
}

impl Solution for Day8 {
    const NAME: &'static str = "Playground";

    fn part_one(&self, inp: &str) -> Self::OutputP1 {
        let (boxes, edges) = Self::parse_input(inp);
        let mut sets = DisjointSet::new(&boxes);

        for (box1, box2) in edges.iter().take(1000) {
            sets.union(box1, box2);
        }

        sets
            .get_sizes()
            .values()
            .sorted_unstable_by(|a, b| b.cmp(a))
            .take(3)
            .product()
    }

    fn part_two(&self, inp: &str) -> Self::OutputP2 {
        let (boxes, edges) = Self::parse_input(inp);
        let mut sets = DisjointSet::new(&boxes);

        let mut last_box1 = None;
        let mut last_box2 = None;

        for (box1, box2) in &edges {
            if sets.count() == 1 {
                break;
            }

            last_box1 = Some(box1);
            last_box2 = Some(box2);
            sets.union(box1, box2);
        }

        last_box1.unwrap()[0] * last_box2.unwrap()[0]
    }

    fn run(&self, inp: String) {
        let p1 = self.part_one(&inp);
        let p2 = self.part_two(&inp);

        println!("Part 1: {p1}");
        println!("Part 2: {p2}");

        assert_eq!(p1, 54_180);
        assert_eq!(p2, 25_325_968);
    }
}

fn main() {
    aoc_2025::run_day(8, &Day8);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() { main(); }
}